wp-sw-cache/lib/service-worker.js

```diff
@@ -1,103 +1,155 @@
+  var wpSwCache = self.wpSwCache = {
+    // A { url: hash } object which will be used to populate cache
+    // A changed url object will ensure the SW is downloaded when URLs change
+    urls: $urls,
+    // Instance of localForage to save urls and hashes to see if anything has changed
+    storage: localforage.createInstance({ name: storageKey }),
+    // Name of the cache the plugin will use
+    cacheName: storageKey,
+    // Method to cleanse a URL before comparing and caching
+    normalizeAndAnonymize: function(request) {
+      var url = new URL(request.url);
+      if (url.origin !== location.origin) {
+        return request;
+      }
 
-  self.addEventListener('install', function(event) {
-    // Perform install step:  loading each required file into cache
-    event.waitUntil(
-      caches.open(CACHE_NAME)
-        .then(function(cache) {
-          // Add all offline dependencies to the cache
-          if (debug) {
-            console.log('[install] Caches opened, adding all core components to cache');
+      url.search = '';
+      url.fragment = '';
+      return new Request(url, {
+        method: request.method,
+        headers: request.headers,
+        mode: 'no-cors',
+        credentials: request.credentials,
+        cache: request.cache,
+        redirect: request.redirect,
+        referrer: request.referrer,
+        integrity: request.integrity
+      });
+    },
+    // Detect if a URL should be cacheable and in the desired URL list
+    shouldBeHandled: function(request) {
+      return request.method === 'GET' && (request.url in this.urls);
+    },
+    // Adds URLs to cache and localForage if the file has changed or needs to be added
+    update: function() {
+      // For every URL (file) the user wants cached...
+      return Promise.all(Object.keys(this.urls).map(url => {
+        var hash = this.urls[url];
+
+        // ... get its hash from storage ...
+        return this.storage.getItem(url).then(value => {
+          // ... and if nothing has changed, just move on to the next URL
+          if(value === hash) {
+            this.log('[update] Hash unchanged, doing nothing: ', url);
+            return Promise.resolve();
           }
-          return cache.addAll(CACHE_FILES);
-        })
-        .then(function() {
-          if (debug) {
-            console.log('[install] All required resources have been cached, we\'re good!');
+          else {
+            this.log('[update] Hash changed or new URL, adding to cache "' + this.cacheName +'" : ' , url);
+            // ... Add the new/updated URL and its response to cache ...
+            return self.caches.open(this.cacheName).then(cache => {
+              return cache.add(url).then(() => {
+                // ... and once it's successful add its hash to storage
+                return this.storage.setItem(url, hash).catch(e => {
+                  this.warn('[update] error: ', e)
+                });
+              })
+              .catch(e => {
```

Comment:
You could remove this and the following `catch` and only have a single one at the end of the promise chain.